src/fcmp_pp/fcmp_pp_rust/src/lib.rs

```diff
@@ -0,0 +1,228 @@
+use ciphersuite::{
+    group::{
+        ff::{Field, PrimeField},
+        GroupEncoding,
+    },
+    Ciphersuite, Helios, Selene,
+};
+use helioselene::{
+    Field25519 as SeleneScalar, HeliosPoint, HelioseleneField as HeliosScalar, SelenePoint,
+};
+
+use ec_divisors::DivisorCurve;
+use full_chain_membership_proofs::tree::{hash_grow, hash_trim};
+
+use monero_fcmp_plus_plus::{HELIOS_HASH_INIT, SELENE_HASH_INIT, HELIOS_GENERATORS, SELENE_GENERATORS};
+
+// TODO: Use a macro to de-duplicate some of of this code
+
+#[no_mangle]
+pub extern "C" fn helios_hash_init_point() -> HeliosPoint {
+    HELIOS_HASH_INIT()
+}
+
+#[no_mangle]
+pub extern "C" fn selene_hash_init_point() -> SelenePoint {
+    SELENE_HASH_INIT()
+}
+
+fn c_u8_32(bytes: [u8; 32]) -> *const u8 {
+    let arr_ptr = Box::into_raw(Box::new(bytes));
+    arr_ptr as *const u8
+}
+
+#[no_mangle]
+pub extern "C" fn helios_scalar_to_bytes(helios_scalar: HeliosScalar) -> *const u8 {
+    c_u8_32(helios_scalar.to_repr())
+}
+
+#[no_mangle]
+pub extern "C" fn selene_scalar_to_bytes(selene_scalar: SeleneScalar) -> *const u8 {
+    c_u8_32(selene_scalar.to_repr())
+}
+
+#[no_mangle]
+pub extern "C" fn helios_point_to_bytes(helios_point: HeliosPoint) -> *const u8 {
+    c_u8_32(helios_point.to_bytes())
+}
+
+#[no_mangle]
+pub extern "C" fn selene_point_to_bytes(selene_point: SelenePoint) -> *const u8 {
+    c_u8_32(selene_point.to_bytes())
+}
+
+#[allow(clippy::not_unsafe_ptr_arg_deref)]
+#[no_mangle]
+pub extern "C" fn helios_point_from_bytes(helios_point: *const u8) -> HeliosPoint {
+    let mut helios_point = unsafe { core::slice::from_raw_parts(helios_point, 32) };
+    // TODO: Return an error here (instead of unwrapping)
```

Comment:
Another thing to mark as `\TODO` is how to handle Rust stack unwinds.  An unwind has undefined behavior once it hits an FFI boundary. There's a hidden (and very un-rust) utility called [`catch_unwind`](https://doc.rust-lang.org/std/panic/fn.catch_unwind.html) that we may have to deploy in some of these functions to keep information details from leaking.

Unfortunately, I'm not really certain on what you do once you catch the unwind. The Rust code is an invalid state which means - immediate abort? Some quick Google searches didn't yield much insight as to what the standard practice is at the FFI boundary.

Reply:
Does `libunwind` allow "pausing" the winding such that it won't be undefined to call the FFI, but just loses the information?

Reply:
Rust will abort the process here if it was to panic: https://doc.rust-lang.org/nomicon/ffi.html#panic-can-be-stopped-at-an-abi-boundary

We could use `catch_unwind` to prevent the abort, however we are also setting `panic = "abort"` in the `Cargo.toml` which will just immediately abort on any panic as well.

Reply:
@Boog900  I missed that. That's probably the only thing that can be done on panic. I don't even know how/why `catch_unwind` is allowed, seems like it ruins tons of invariants if used to do something (other than abort).

Reply:
I figure it makes sense to catch and handle errors, while panics should reasonably abort (and panics should never trigger unless the code is broken). Sounds like you'd agree?

Reply:
Yes. If there's an error returned on the call stack, you can probably bubble that up through the FFI, otherwise you basically have to abort. Rust code really isn't written to be "exception-safe", so lots of invariants would break if an unwind occurs.